File: java/src/main/java/fyi/oxide/pdf/PdfPolicy.java
/*
 * Copyright 2025-2026 Yury Fedoseev and pdf_oxide contributors.
 * Licensed under MIT OR Apache-2.0.
 */
package fyi.oxide.pdf;

import fyi.oxide.pdf.internal.NativeLoader;
import fyi.oxide.pdf.policy.PolicyMode;

/**
 * Process-global crypto-governance policy (v0.3.50 #230).
 *
 * <p>Selects which cryptographic algorithms are accepted for reads
 * and writes. Composes with the build-time feature flags
 * ({@code legacy-crypto}, {@code fips}) — if a build is missing
 * {@code legacy-crypto} then {@link PolicyMode#COMPAT} can't enable
 * RC4/MD5-KDF (the algorithm isn't compiled in regardless of policy).
 *
 * <p><b>Set-once semantics.</b> pdf_oxide installs the policy at
 * most once per process: call {@link #set} <b>before</b> any other
 * pdf_oxide operation (including {@link #current}). A second
 * {@link #set} call — or one after any document has been opened
 * — throws {@link fyi.oxide.pdf.exception.PdfException} with a
 * message containing {@code "already set"}. This is deliberate: a
 * runtime policy downgrade would be a security attack vector.
 *
 * <p>If no explicit {@link #set} call is made, {@link #current} (or
 * any first crypto access) lazily installs {@link PolicyMode#COMPAT}.
 */
public final class PdfPolicy {

    static {
        NativeLoader.ensureLoaded();
    }

    private PdfPolicy() {
        // Static-only.
    }

    /** @return the process-current policy mode. */
    public static PolicyMode current() {
        return ORDINAL_TO_MODE[nativeCurrentOrdinal()];
    }

    /** Set the process-global policy mode. */
    public static void set(PolicyMode mode) {
        java.util.Objects.requireNonNull(mode, "mode");
        nativeSetByOrdinal(mode.ordinal());
    }

    /**
     * Lookup table indexed by the {@link PolicyMode} ordinal — must
     * stay in sync with the constants in
     * {@code pdf_oxide_jni/src/policy.rs}. Validated by a unit test
     * that checks the enum constant order.
     */
    private static final PolicyMode[] ORDINAL_TO_MODE = PolicyMode.values();

    private static native int nativeCurrentOrdinal();

    private static native void nativeSetByOrdinal(int ordinal);

    /** Preset: accept all algorithms (RC4, MD5-KDF). Default mode. */
    public static PolicyMode compat() {
        return PolicyMode.COMPAT;
    }
    /** Preset: reject legacy algorithms. */
    public static PolicyMode strict() {
        return PolicyMode.STRICT;
    }
    /** Preset: FIPS 140-3 only. Requires the {@code fips} build feature. */
    public static PolicyMode fipsStrict() {
        return PolicyMode.FIPS_STRICT;
    }
}

File: ruby/lib/pdf_oxide/pdf_policy.rb
# frozen_string_literal: true

module PdfOxide
  # Process-global crypto-governance policy (v0.3.50 #230).
  #
  # Mirrors `fyi.oxide.pdf.PdfPolicy`.  Selects which cryptographic
  # algorithms are accepted for reads and writes.  Composes with the
  # build-time feature flags (`legacy-crypto`, `fips`) — if a build
  # lacks `legacy-crypto`, COMPAT can't enable RC4/MD5-KDF regardless
  # of policy.
  #
  # **Set-once semantics.**  pdf_oxide installs the policy at most
  # once per process: call {.set} **before** any other pdf_oxide
  # operation.  A second `.set` call — or one after any document has
  # been opened — raises with a message containing "already set".
  module PdfPolicy
    # Policy modes (mirrors Java's `PolicyMode` enum).
    MODES = { compat: 0, strict: 1, fips_strict: 2 }.freeze
    ORDINAL_TO_MODE = MODES.invert.freeze

    module_function

    # @return [Symbol] the current process policy mode (:compat / :strict / :fips_strict).
    def current
      ord = Bindings.pdf_oxide_policy_current_ordinal if Bindings.respond_to?(:pdf_oxide_policy_current_ordinal)
      ord ||= 0 # default COMPAT if accessor not exposed in this build
      ORDINAL_TO_MODE.fetch(ord, :compat)
    rescue ::FFI::NotFoundError
      :compat
    end

    # Set the process-global policy mode.  Call before any other
    # pdf_oxide operation.
    # @param mode [Symbol]
    # @raise [InternalError] policy was already set.
    def set(mode)
      ordinal = MODES.fetch(mode) do
        raise ::PdfOxide::ArgumentError, "mode must be one of #{MODES.keys.inspect}, got #{mode.inspect}"
      end
      raise UnsupportedFeatureError, 'policy not supported by this cdylib build' \
        unless Bindings.respond_to?(:pdf_oxide_policy_set_by_ordinal)

      rc = Bindings.pdf_oxide_policy_set_by_ordinal(ordinal)
      raise InternalError, 'policy already set' if rc != 0

      mode
    end

    # @return [Symbol] :compat preset (accept all algorithms).
    def compat
      :compat
    end

    # @return [Symbol] :strict preset (reject legacy algorithms).
    def strict
      :strict
    end

    # @return [Symbol] :fips_strict preset (FIPS 140-3 only).
    def fips_strict
      :fips_strict
    end
  end
end
